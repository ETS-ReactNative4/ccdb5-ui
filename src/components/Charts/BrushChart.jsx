import './BrushChart.less'
import * as d3 from 'd3'
import { brush } from 'britecharts'
import { changeDates } from '../../actions/filter'
import { connect } from 'react-redux'
import moment from 'moment'
import React from 'react'

export class BrushChart extends React.Component {
  constructor( props ) {
    super( props )

    // This binding is necessary to make `this` work in the callback
    // https://facebook.github.io/react/docs/handling-events.html
    this._brushEnd = this._brushEnd.bind( this )
  }

  componentDidMount() {
    this._redrawChart()
  }

  componentDidUpdate( prevProps ) {
    // Typical usage (don't forget to compare props):
    if ( JSON.stringify( this.props ) !== JSON.stringify( prevProps ) ) {
      this._redrawChart()
    }
  }

  _brushEnd( brushExtent ) {
    const format = d3.timeFormat( '%m/%d/%Y' )

    let from = brushExtent[0]
    let through = brushExtent[1]
    d3.select( '.js-start-date' ).text( format( from ) )
    d3.select( '.js-end-date' ).text( format( through ) )
    d3.select( '.js-date-range' ).classed( 'is-hidden', false )
    d3.select( 'rect.handle.handle--e, rect.handle.handle--w' )
      .attr( 'y', '1' )
    from = moment( from, 'MM-DD-YYYY' )
    through = moment( through, 'MM-DD-YYYY' )

    // let dateRange;

    // if ( !brushExtent[0] || !brushExtent[1] ) {
    //   dateRange = {
    //     from: defaultDateRange.min,
    //     to: defaultDateRange.max
    //   };
    // } else {
    //   dateRange = {
    //     from: formatDateModel( brushExtent[0] ),
    //     to: formatDateModel( brushExtent[1] )
    //   }
    // }
    this.props.changeDates( from, through )
  }

  _redrawChart() {
    const chartID = '#brush-chart'
    const container = d3.select( chartID )
    d3.select( chartID + ' .brush-chart' ).remove()
    const containerWidth =
      container.node() ?
        container.node().getBoundingClientRect().width :
        false
    const brushChart = brush()

    brushChart.margin( {
      left: 30,
      right: 10,
      top: 10,
      bottom: 40
    } )
      .width( containerWidth )
      .height( 70 )
      .on( 'customBrushEnd', this._brushEnd )

    if ( this.props.brushDateData.length ) {
      // console.log( this.props.brushDateData )
      container.datum( this.props.brushDateData ).call( brushChart )
      brushChart.dateRange( this.props.dateRange )
    }


  }

  _getDateRange() {
    // const dr = $scope.dateRange
    // const min = isDateGreater( defaultDateRange.min, dr.from ) ?
    //   defaultDateRange.min : dr.from
    //
    // const max = isDateGreater( defaultDateRange.max, dr.to ) ?
    //   dr.to : defaultDateRange.max

    // return [
    //   this.props,
    // ]
  }

  render() {
    return (
      <div id="brush-chart">
      </div>
    )
  }
}

export const mapStateToProps = state => ( {
  brushDateData: state.trends.results.dateRangeBrush,
  dateRange: [ state.query.date_received_min, state.query.date_received_max ],
  lens: state.query.lens
} )

export const mapDispatchToProps = dispatch => ( {
  changeDates: ( from, through ) => {
    dispatch( changeDates( 'date_received', from, through ) )
  }
} )

export default connect( mapStateToProps, mapDispatchToProps )( BrushChart )
