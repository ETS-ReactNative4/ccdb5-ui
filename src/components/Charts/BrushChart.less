@import (less) "../../css/base.less";

#brush-chart {
  .brush-chart {
    .x-axis-group {
      display: none;
    }

    .chart-group {
      path {
        //fill: @green-60;
        opacity: 0.5;
      }
    }

    .brush-group {
      rect {
        &.selection.brush-rect {
          //fill-opacity: 0.2;
          stroke: none;
        }

        &.handle.brush-rect {
          fill: @gray-10;
          fill-opacity: .3;
          stroke: @gray-40;
          stroke-opacity: 1;
          stroke-width: 1px;
          width: 10px;
        }
      }
    }
  }

  &.overview-brush-chart {
    .chart-group {
      path {
        fill: @green-60;
      }
    }
  }

  .brush-group {
    rect {
      &.selection.brush-rect {
        fill-opacity: 0.2;
        stroke: none;
      }
    }
  }

  .chart-group {
    path {
      fill: #c3c4c4;
    }
  }
}
