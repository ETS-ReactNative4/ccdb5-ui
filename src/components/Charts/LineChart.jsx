import './LineChart.less'
import * as d3 from 'd3'
import { line, tooltip } from 'britecharts'
import { connect } from 'react-redux'
import { getTipDate } from '../../utils/chart'
import React from 'react'

export class LineChart extends React.Component {
  componentDidUpdate() {
    this._redrawChart()
  }

  _redrawChart() {
    const chartID = '#line-chart'
    const container = d3.select( chartID )
    const containerWidth =
      container.node() ?
        container.node().getBoundingClientRect().width :
        false
    d3.select( chartID + ' .line-chart' ).remove()
    const lineChart = line()
    const tip = tooltip()

    const dateInterval = this.props.dateInterval
    // will be Start Date - Date...

    lineChart.margin( { left: 50, right: 10, top: 10, bottom: 40 } )
      .initializeVerticalMarker( true )
      .isAnimated( true )
      .tooltipThreshold( 1 )
      .grid( 'horizontal' )
      .aspectRatio( 0.5 )
      .width( containerWidth )
      .dateLabel( 'date' )
      .colorSchema( [ 'green' ] )
      .on( 'customMouseOver', tip.show )
      .on( 'customMouseMove', function(dataPoint, topicColorMap, dataPointXPosition) {
        tip.title( getTipDate( dataPoint.date, dateInterval ) )
        tip.update( dataPoint, topicColorMap, dataPointXPosition )
      } )
      .on( 'customMouseOut', tip.hide )

    container.datum( this.props.data ).call( lineChart )
    const tooltipContainer =
      d3.select( chartID + ' .metadata-group .vertical-marker-container' )
    tooltipContainer.datum( [] ).call( tip )
  }

  render() {
    return (
      <div>
        <h2>{ this.props.title }</h2>
        <div id="line-chart">
        </div>
      </div>
    )
  }
}

export const mapStateToProps = state => ( {
  data: state.trends.results.dateRangeLine,
  dateInterval: state.query.dateInterval
} )

export default connect( mapStateToProps )( LineChart )
