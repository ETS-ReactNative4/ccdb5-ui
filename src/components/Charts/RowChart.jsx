/* eslint complexity: ["error", 5] */

import './RowChart.less'
import * as d3 from 'd3'
import { miniTooltip, row } from 'britecharts'
import { connect } from 'react-redux'
import { hashObject } from '../../utils'
import { max } from 'd3-array'
import PropTypes from 'prop-types'
import React from 'react'
import { toggleTrend } from '../../actions/trends'

export class RowChart extends React.Component {
  _getHeight( numRows ) {
    return numRows === 1 ? 100 : numRows * 60
  }

  _wrapText( text, width ) {
    // ignore test coverage since this is code borrowed from d3 mbostock
    // text wrapping functions
    /* istanbul ignore next */
    // eslint-disable-next-line complexity
    text.each( function() {
      const innerText = d3.select( this )
      if ( innerText.node().children && innerText.node().children.length > 0 ) {
        // assuming its already split up
        return
      }
      const words = innerText.text().split( /\s+/ ).reverse(),
            // ems
            lineHeight = 1.1,
            y = innerText.attr( 'y' ) || 0,
            dy = parseFloat( innerText.attr( 'dy' ) || 0 )

      let word,
          line = [],
          lineNumber = 0,
          tspan = innerText.text( null ).append( 'tspan' ).attr( 'x', 0 )
            .attr( 'y', y ).attr( 'dy', dy + 'em' )

      // eslint-disable-next-line no-cond-assign
      while ( word = words.pop() ) {
        line.push( word )
        tspan.text( line.join( ' ' ) )
        if ( tspan.node().getComputedTextLength() > width ) {
          line.pop()
          tspan.text( line.join( ' ' ) )
          line = [ word ]
          tspan = innerText.append( 'tspan' )
            .attr( 'x', 0 )
            .attr( 'y', y )
            // eslint-disable-next-line no-mixed-operators
            .attr( 'dy', ++lineNumber * lineHeight + dy + 'em' )
            .text( word )
        }
      }
    } )
  }


  componentDidUpdate( prevProps ) {
    const props = this.props
    if( hashObject( prevProps ) !== hashObject( props ) ) {
      this._redrawChart()
    }
  }
  // --------------------------------------------------------------------------
  // Event Handlers
  // eslint-disable-next-line complexity
  _redrawChart() {
    const { colorScheme, data, id, printMode, toggleRow, total } = this.props
    if ( !data || !data.length ) {
      return
    }

    const tooltip = miniTooltip()
    tooltip.valueFormatter( value => value.toLocaleString() + ' complaints' )

    const rows = data
    const ratio = total / max( rows, o => o.value )
    const chartID = '#row-chart-' + id
    d3.selectAll( chartID + ' .row-chart' ).remove()
    const rowContainer = d3.select( chartID )
    const width = printMode ? 750 :
      rowContainer.node().getBoundingClientRect().width
    const height = this._getHeight( rows.length )
    const chart = row()
    const marginLeft = width / 3

    // tweak to make the chart full width at desktop
    // add space at narrow width
    const marginRight = width < 600 ? 20 : -80
    chart.margin( {
      left: marginLeft,
      right: marginRight,
      // left: 200,
      // right: 50,
      top: 20,
      bottom: 10
    } )
      .colorSchema( colorScheme )
      .backgroundColor( '#f7f8f9' )
      // .enableYAxisRight(true) // for trends to show deltas
      .enableLabels( true )
      .labelsTotalCount( total.toLocaleString() )
      .labelsNumberFormat( ',d' )
      .outerPadding( 0.1 )
      .percentageAxisToMaxRatio( ratio )
      .yAxisLineWrapLimit( 2 )
      .yAxisPaddingBetweenChart( 20 )
      .width( width )
      .wrapLabels( true )
      .height( height )
      .on( 'customMouseOver', tooltip.show )
      .on( 'customMouseMove', tooltip.update )
      .on( 'customMouseOut', tooltip.hide )

    rowContainer.datum( rows ).call( chart )
    const tooltipContainer = d3.selectAll( chartID + ' .row-chart .metadata-group' )
    tooltipContainer.datum( [] ).call( tooltip );
    this._wrapText( d3.select( chartID ).selectAll( '.tick text' ), marginLeft )

    rowContainer
      .selectAll( '.y-axis-group .tick' )
      .on( 'click', toggleRow )
  }

  render() {
    return (
      <div className="row-chart-section">
        <h3>{ this.props.title }</h3>
        <div id={ 'row-chart-' + this.props.id }>
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ( {
  toggleRow: selectedState => {
    dispatch( toggleTrend( selectedState ) )
  }
} )

export const mapStateToProps = ( state, ownProps ) => {
  const { printMode, width } = state.view
  return {
    printMode,
    total: state.aggs.total,
    width
  }
}

RowChart.propTypes = {
  id: PropTypes.string.isRequired,
  colorScheme: PropTypes.oneOfType( [
    PropTypes.array,
    PropTypes.bool
  ] ).isRequired,
  data: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired
}

export default connect( mapStateToProps, mapDispatchToProps )( RowChart )
