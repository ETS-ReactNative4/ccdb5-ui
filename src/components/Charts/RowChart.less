@import (less) "../../css/base.less";

.row-chart-section {
  margin-left: @grid_gutter-width;
  margin-top: @grid_gutter-width/1.5;
}
.row-chart {
  .bg-hover, .visibility, line, path.domain, .x-axis-group {
    display: none;
  }
  text {
    font-size: 16px;
    &.adjust-upwards {
      transform: translateY(-12px);
    }
    &.hidden {
      display:none;
    }
  }
  .view-more-label {
    font-size: 12px;
    text-decoration: underline;
    fill: @pacific;
  }
}
