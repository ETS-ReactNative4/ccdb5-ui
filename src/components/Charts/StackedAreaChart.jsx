import './StackedAreaChart.less'
import * as colors from '../../constants/colors'
import * as d3 from 'd3'
import { connect } from 'react-redux'
import { getLastDate } from '../../utils/chart'
import { hashObject } from '../../utils'
import { isDateEqual } from '../../utils/formatDate'
import React from 'react'
import { stackedArea } from 'britecharts'
import { updateTrendsTooltip } from '../../actions/trends'


export class StackedAreaChart extends React.Component {
  constructor( props ) {
    super( props )
    this._updateTooltip = this._updateTooltip.bind( this )
  }

  componentDidMount() {
    this._redrawChart()
  }

  componentDidUpdate( prevProps ) {
    const props = this.props
    if ( hashObject( prevProps.data ) !== hashObject( props.data ) ||
      hashObject( prevProps.width ) !== hashObject( props.width ) ||
      hashObject( prevProps.printMode ) !== hashObject( props.printMode ) ) {
      this._redrawChart()
    }
  }

  _updateTooltip( point ) {
    if ( !isDateEqual( this.props.tooltip.date, point.date ) ) {
      this.props.tooltipUpdated( {
        date: point.date,
        dateRange: this.props.dateRange,
        interval: this.props.interval,
        values: point.values
      } )
    }
  }

  _chartWidth( chartID ) {
    const { printMode } = this.props
    if ( printMode ) {
      return 540
    }
    const container = d3.select( chartID )
    return container.node().getBoundingClientRect().width
  }


  _redrawChart() {
    const {
      colorMap, data, dateRange, interval, lastDate
    } = this.props
    if ( !data || !data.length ) {
      return
    }

    const chartID = '#stacked-area-chart'
    const container = d3.select( chartID )
    const width = this._chartWidth( chartID )
    d3.select( chartID + ' .stacked-area' ).remove()
    const stackedAreaChart = stackedArea()
    const colorData = data.filter(
      item => item.name !== 'Other'
    )
    const colorScheme = [ ...new Set( colorData.map( item => item.name ) ) ]
      .map( o => colorMap[o] )
    colorScheme.push( colors.DataLens[10] )

    stackedAreaChart.margin( { left: 70, right: 10, top: 10, bottom: 40 } )
      .areaCurve( 'linear' )
      .initializeVerticalMarker( true )
      .isAnimated( false )
      .tooltipThreshold( 1 )
      .grid( 'horizontal' )
      .aspectRatio( 0.5 )
      .width( width )
      .dateLabel( 'date' )
      .colorSchema( colorScheme )
      .on( 'customMouseMove', this._updateTooltip )

    container.datum( data ).call( stackedAreaChart )

    const config = {
      dateRange,
      interval,
      lastDate
    }

    this.props.tooltipUpdated( getLastDate( data, config ) )
  }

  render() {
    return (
      <div>
        <div id="stacked-area-chart">
        </div>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ( {
  tooltipUpdated: selectedState => {
    // Analytics.sendEvent(
    //   Analytics.getDataLayerOptions( 'Trend Event: add',
    //     selectedState.abbr, )
    // )
    dispatch( updateTrendsTooltip( selectedState ) )
  }
} )

export const mapStateToProps = state => ( {
  colorMap: state.trends.colorMap,
  data: state.trends.results.dateRangeArea,
  dateRange:  {
    from: state.query.date_received_min,
    to: state.query.date_received_max
  },
  interval: state.query.dateInterval,
  lastDate: state.trends.lastDate,
  lens: state.trends.lens,
  printMode: state.view.printMode,
  tooltip: state.trends.tooltip,
  width: state.view.width
} )

export default connect( mapStateToProps,
  mapDispatchToProps )( StackedAreaChart )
