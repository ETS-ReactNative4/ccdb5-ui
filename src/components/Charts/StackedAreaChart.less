@import (less) "../../css/base.less";

#stacked-area-chart {
  .stacked-area {
    .y-axis-group {
      .domain {
        display: none;
      }
    }
  }

}

.chart-wrapper {
  p {
    font-size: 12px;
    font-weight: 600;
    color: @gray;

    &.y-axis-label {
      margin-left: @gutter-normal;
    }

    &.x-axis-label {
      margin-left: 45%;
    }
  }
}
