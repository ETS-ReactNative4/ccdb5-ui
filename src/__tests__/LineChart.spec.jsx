import configureMockStore from 'redux-mock-store'
import { mapStateToProps, LineChart } from '../components/Charts/LineChart'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import thunk from 'redux-thunk'

// this is how you override and mock an imported constructor
jest.mock( 'britecharts', () => {
  const props = [
    'line', 'margin', 'backgroundColor', 'colorSchema', 'enableLabels',
    'labelsSize', 'labelsTotalCount', 'labelsNumberFormat', 'outerPadding',
    'percentageAxisToMaxRatio', 'yAxisLineWrapLimit', 'grid', 'dateLabel',
    'initializeVerticalMarker', 'yAxisPaddingBetweenChart', 'width', 'on',
    'wrapLabels', 'height', 'isAnimated', 'tooltipThreshold', 'aspectRatio',
    // tooltip specifics
    'tooltip', 'title', 'update'
  ]

  const mock = {}

  for ( let i = 0; i < props.length; i++ ) {
    const propName = props[i]
    mock[propName] = jest.fn().mockImplementation( () => {
      return mock
    } )
  }

  return mock
} )

jest.mock( 'd3', () => {
  const props = [
    'select', 'each', 'node', 'getBoundingClientRect', 'width', 'datum', 'call',
    'remove', 'selectAll'
  ]

  const mock = {}

  for ( let i = 0; i < props.length; i++ ) {
    const propName = props[i]
    mock[propName] = jest.fn().mockImplementation( () => {
      return mock
    } )
  }

  // set narrow width value for 100% test coverage
  mock.width = 100

  return mock
} )

function setupSnapshot() {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore( middlewares )
  const store = mockStore( {
    map: {}
  } )

  return renderer.create(
    <Provider store={ store }>
      <LineChart aggtype={'foo'}/>
    </Provider>
  )
}

describe( 'component: LineChart', () => {
  describe( 'initial state', () => {
    it( 'renders without crashing', () => {
      const target = setupSnapshot()
      let tree = target.toJSON()
      expect( tree ).toMatchSnapshot()
    } )
  } )

  describe( 'componentDidUpdate', () => {
    let mapDiv

    beforeEach( () => {
      mapDiv = document.createElement( 'div' )
      mapDiv.setAttribute( 'id', 'line-chart-foo' )
      window.domNode = mapDiv
      document.body.appendChild( mapDiv )
    } )

    afterEach( () => {
      const div = document.getElementById( 'line-chart-foo' )
      if ( div ) {
        document.body.removeChild( div )
      }
      jest.clearAllMocks()
    } )

    it( 'does nothing when no data', () => {
      const target = shallow( <LineChart data={ [] } aggtype={'foo'}/> )
      target._redrawChart = jest.fn()
      target.setProps( { data: [] } )
      expect(  target._redrawChart ).toHaveBeenCalledTimes( 0 )
    } )

    it( 'trigger a new update when data changes', () => {
      const target = shallow( <LineChart data={ [ 23, 4, 3 ] } aggtype={'foo'} total={1000}/> )
      target._redrawChart = jest.fn()
      const sp = jest.spyOn(target.instance(), '_redrawChart')
      target.setProps( { data: [ 2, 5 ] } )
      expect( sp ).toHaveBeenCalledTimes( 1 )
    } )

    it( 'trigger a new update when printMode changes', () => {
      const target = shallow( <LineChart data={ [ 23, 4, 3 ] }
                                        aggtype={'foo'} total={1000}
                                        printMode={'false'}
      /> )
      target._redrawChart = jest.fn()
      const sp = jest.spyOn(target.instance(), '_redrawChart')
      target.setProps( { printMode: true } )
      expect( sp ).toHaveBeenCalledTimes( 1 )
    } )

    it( 'trigger a new update when width changes', () => {
      const target = shallow( <LineChart data={ [ 23, 4, 3 ] }
                                        aggtype={'foo'} total={1000}
                                        printMode={'false'}
                                        width={1000}
      /> )
      target._redrawChart = jest.fn()
      const sp = jest.spyOn(target.instance(), '_redrawChart')
      target.setProps( { width: 600 } )
      expect( sp ).toHaveBeenCalledTimes( 1 )
    } )
  } )

  describe( 'mapStateToProps', () => {
    it( 'maps state and props', () => {
      const state = {
        aggs: {
          total: 100
        },
        map: {
          results: {
            baz: [ 1, 2, 3 ]
          }
        },
        query: {
          baz: [ 1, 2, 3 ],
          dateInterval: 'Month'
        },
        trends: {
          results: {
            dateRangeArea: []
          }
        },
        view: {
          printMode: false
        }
      }
      const ownProps = {
        aggtype: 'baz'
      }
      let actual = mapStateToProps( state, ownProps )
      expect( actual ).toEqual( {
        data: [],
        dateInterval: 'Month'
      } )
    } )
  } )
} )
