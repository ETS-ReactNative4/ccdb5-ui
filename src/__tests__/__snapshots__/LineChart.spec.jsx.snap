// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: LineChart initial state renders without crashing 1`] = `
<div>
  <h2 />
  <div
    id="line-chart"
  />
</div>
`;
